### DISCUSSION TITLE
Getting the exact run command that an configuration produces

### DISCUSSION THREAD

**Discussion by invertedEcho** [zengkid/SmartTomcat#98] [General]
Hello,

I'm currently working with intellij ultimate und would like to be able to run the exact same run config on for example neovim.
For that, I would need the run command that gets runned by the tomcat local server configuration, so I can just run it in an terminal as an example.
Is something like that possible?

**Comment by yuezk** [comment]
I don't quite understand your problem. How can you run IDEA with neovim? Can you describe the workflow in depth?

**Comment by invertedEcho** [comment]
I would like to switch from idea to nvim. For that, I would like to get the run command that gets executed when running the tomcat run config, so I can just run it in an terminal to be independen of idea. I know that it just executes the startup.sh script but thats prob not all, thats why i want to get the run command that my tomcat run config produces.

**Reply by yuezk** [reply]
The run command for Smart Tomcat is totally different. It didn't call `startup.sh` or `catalina.sh`, because I didn't find a way to specify the debugger port.

Instead, we build a `java` command (see below), which is the same as the command built by `startup.sh` or `catalina.sh`.

```
/Users/zyue/Library/Java/JavaVirtualMachines/azul-11.0.17/Contents/Home/bin/java \
  -agentlib:jdwp=transport=dt_socket,address=127.0.0.1:61305,suspend=y,server=n \
  -Dcatalina.home=/Users/zyue/work/tools/apache-tomcat-9.0.58 \
  -Dcatalina.base=/Users/zyue/.SmartTomcat/web-dossier/web-dossier.LibraryRest.main \
  -Djava.io.tmpdir=/Users/zyue/.SmartTomcat/web-dossier/web-dossier.LibraryRest.main/temp \
  -Djava.util.logging.config.file=/Users/zyue/.SmartTomcat/web-dossier/web-dossier.LibraryRest.main/conf/logging.properties \
  -Djava.util.logging.manager=org.apache.juli.ClassLoaderLogManager \
  -javaagent:/Users/zyue/Library/Caches/JetBrains/IntelliJIdea2022.2/captureAgent/debugger-agent.jar=file:/private/var/folders/9s/bkzz5_792tb58ptl8p9j7pwd75fbzw/T/capture.props \
  -Dfile.encoding=UTF-8 \
  -classpath "/Users/zyue/work/tools/apache-tomcat-9.0.58/bin/bootstrap.jar:/Users/zyue/work/tools/apache-tomcat-9.0.58/bin/tomcat-juli.jar:/Users/zyue/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/222.4345.14/IntelliJ IDEA.app/Contents/lib/idea_rt.jar" \
  org.apache.catalina.startup.Bootstrap start
```

**Reply by invertedEcho** [reply]
Alright, will try that out and edit the command accordingly to my paths.
I also need to build an artifact, a WAR, can I also find out how exactly that is being built and where the war is being placed so tomcat finds it?

**Reply by yuezk** [reply]
This plugin won't build the WAR file, it collects the dependencies and creates a context.xml under the `~/.SmartTomcat/<project_name>/<module_name>/conf/Catalina` folder.

**Reply by invertedEcho** [reply]
Hmm, I don't have that ~/.SmartTomcat folder. Any ideas on why?

**Reply by yuezk** [reply]
Have you ever run SmartTomcat in IDEA?

**Reply by invertedEcho** [reply]
Yes, I have. I think its because I have tomcat installed in ~/.local/bin/. I do have conf/Catalina/localhost, but nothing is in there.

**Reply by yuezk** [reply]
If you launch Smart Tomcat in IDEA, you could see the logs in the Console. Pay attention to the value of `CATALINA_BASE`.

<img width="1631" alt="image" src="https://user-images.githubusercontent.com/3297602/199175064-18856804-fba6-4fc1-853d-2b4434bd552f.png">

**Reply by invertedEcho** [reply]
This is my CATALINA_BASE
CATALINA_BASE: /home/user/.cache/JetBrains/IntelliJIdea2022.2/tomcat/2cea9274-6af6-4e31-93a8-89a626fe6688

**Reply by yuezk** [reply]
I think you are using the official Tomcat plugin of the Ultimate version.

**Reply by invertedEcho** [reply]
Sorry for the late answer, forgot about this. Yes, you were right, I mistook the plugin, I am using the official TomCat plugin of the ultimate version, I thought it was the SmartTomcat plugin because I had it installed but i dont use it. By any chance, do you know how to find out how exactly a specific artifact is being built, like the command of it?

**Reply by yuezk** [reply]
The official Tomcat plugin will build an artifact and then deploy that artifact to the Tomcat server. But SmartTomcat doesn't do that; instead, it collects all the classpaths and builds a context.xml for running the application.

The official plugin can read the artifacts and build them programmatically. There could be some plugin APIs to achieve this, but I don't know the details.

<img width="184" alt="image" src="https://user-images.githubusercontent.com/3297602/208365463-0f7d178f-8af3-458d-839f-eb294c8bf123.png">

**Reply by invertedEcho** [reply]
But you don't know how to retreive the build command that gets run when building the artifact provided by IntelliJ ultimate? Thats what I am interested in. I know thats off topic but hoping you may know that

**Reply by yuezk** [reply]
I don’t know that

**Reply by invertedEcho** [reply]
Well, still thank you for trying to help!
